services/subnet_manager/src/main/java/com/futurewei/alcor/subnet/controller/SubnetController.java

```diff
@@ -0,0 +1,241 @@
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static org.springframework.web.bind.annotation.RequestMethod.*;
+
+@RestController
+public class SubnetController {
+
+    @Autowired
+    private SubnetDatabaseService subnetDatabaseService;
+
+    @Autowired
+    private SubnetService subnetService;
+
+    @RequestMapping(
+            method = GET,
+            value = {"/project/{projectid}/subnets/{subnetId}", "v4/{projectid}/subnets/{subnetId}"})
+    public SubnetStateJson getSubnetStateById(@PathVariable String projectid, @PathVariable String subnetId) throws Exception {
+
+        SubnetState subnetState = null;
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(subnetId);
+            RestPreconditionsUtil.verifyResourceFound(projectid);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(subnetId);
+        } catch (ParameterNullOrEmptyException e) {
+            //TODO: REST error code
+            throw new Exception(e);
+        }
+
+        if (subnetState == null) {
+            //TODO: REST error code
+            return new SubnetStateJson();
+        }
+
+        return new SubnetStateJson(subnetState);
+    }
+
+    @RequestMapping(
+            method = POST,
+            value = {"/project/{projectid}/subnets", "v4/{projectid}/subnets"})
+    @ResponseStatus(HttpStatus.CREATED)
+    public SubnetStateJson createSubnetState(@PathVariable String projectid, @RequestBody SubnetStateJson resource) throws Exception {
+        SubnetState subnetState = null;
+        String portId = UUID.randomUUID().toString();
+
+        try {
+            RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectid);
+            RestPreconditionsUtil.verifyResourceNotNull(resource.getSubnet());
+
+            // TODO: Create a verification framework for all resources
+            SubnetState inSubnetState = resource.getSubnet();
+            RestPreconditionsUtil.verifyResourceFound(inSubnetState.getVpcId());
+            RestPreconditionsUtil.populateResourceProjectId(inSubnetState, projectid);
+
+            this.subnetDatabaseService.addSubnet(inSubnetState);
+
+            subnetState = this.subnetDatabaseService.getBySubnetId(inSubnetState.getId());
+            if (subnetState == null) {
+                throw new ResourcePersistenceException();
+            }
+
+            // Verify VPC ID
+            VpcStateJson vpcResponse = this.subnetService.verifyVpcId(projectid, inSubnetState.getVpcId());
+            if (vpcResponse == null) {
+                throw new FallbackException("fallback request");
+            }
+
+            //Prepare Route Rule(IPv4/6) for Subnet
+            RouteWebJson routeResponse = this.subnetService.createRouteRules(inSubnetState.getId(), inSubnetState);
+            if (routeResponse == null) {
+                throw new FallbackException("fallback request");
```

Comment:
I don't think failing to create route rules should trigger fall back. We should retry first.

Reply:
yeah I agree with you, let me think about it